inc/appliance.class.php

```diff
@@ -0,0 +1,602 @@
+   }
+
+
+   /**
+    * Type than could be linked to a Appliance
+    *
+    * @param boolean $all All types  or only allowed ones (defauts to false)
+    *
+    * @return array of types
+   **/
+   static function getTypes($all = false) {
+      global $CFG_GLPI;
+
+      $types = $CFG_GLPI['appliance_types'];
+
+      if ($all) {
+         return $types;
+      }
+
+      // Only allowed types
+      foreach ($types as $key => $type) {
+         if (!($item = getItemForItemtype($type))) {
+            continue;
+         }
+
+         if (!$item->canView()) {
+            unset($types[$key]);
+         }
+      }
+      return $types;
+   }
+
+
+   function getSpecificMassiveActions($checkitem = null) {
+
+      $isadmin = static::canUpdate();
+      $actions = parent::getSpecificMassiveActions($checkitem);
+
+      if ($_SESSION['glpiactiveprofile']['interface'] == 'central') {
+         if ($isadmin) {
+            $actions['Appliance'.MassiveAction::CLASS_ACTION_SEPARATOR.'install'] = _x('button', 'Associate');
+            $actions['Appliance'.MassiveAction::CLASS_ACTION_SEPARATOR.'uninstall'] = _x('button', 'Dissociate');
+         }
+      }
+      return $actions;
+   }
+
+
+   static function showMassiveActionsSubForm(MassiveAction $ma) {
+
+      switch ($ma->getAction()) {
+         case 'add_item':
+            self::dropdownMA([]);
+            echo "&nbsp;". Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+
+         case "install" :
+            Dropdown::showSelectItemFromItemtypes([
+               'items_id_name' => 'item_item',
+               'itemtype_name' => 'typeitem',
+               'itemtypes'     => self::getTypes(true),
+               'checkright'    => true
+            ]);
+            echo Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+
+         case "uninstall" :
+            Dropdown::showSelectItemFromItemtypes([
+               'items_id_name' => 'item_item',
+               'itemtype_name' => 'typeitem',
+               'itemtypes'     => self::getTypes(true),
+               'checkright'    => true
+            ]);
+            echo Html::submit(_x('button', 'Post'), ['name' => 'massiveaction']);
+            return true;
+      }
+      return parent::showMassiveActionsSubForm($ma);
+   }
+
+
+   static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBTM $item, array $ids) {
+
+      $appliance_item = new Appliance_Item();
+
+      switch ($ma->getAction()) {
+         case "add_item":
+            $input = $ma->getInput();
+            foreach ($ids as $id) {
+               $input = [
+                  'appliances_id'   => $input['appliances_id'],
+                  'items_id'        => $id,
+                  'itemtype'        => $item->getType()
+               ];
+               if ($appliance_item->can(-1, CREATE, $input)) {
+                  if ($appliance_item->add($input)) {
```

Comment:
To prevent SQL error when an appliance is already installed, you should check appliance item existence prior to adding it.

```
[2020-04-07 07:39:35] glpisqllog.ERROR: DBmysql::query() in /var/www/glpi/inc/dbmysql.class.php line 305
  *** MySQL query error:
  SQL: INSERT INTO `glpi_appliances_items` (`appliances_id`, `items_id`, `itemtype`) VALUES ('1', '1', 'Computer')
  Error: Duplicate entry '1-1-Computer' for key 'unicity'
  Backtrace :
  inc/dbmysql.class.php:1131                         
  inc/commondbtm.class.php:592                       DBmysql->insert()
  inc/commondbtm.class.php:1129                      CommonDBTM->addToDB()
  inc/appliance.class.php:552                        CommonDBTM->add()
  inc/massiveaction.class.php:1050                   Appliance::processMassiveActionsForOneItemtype()
  inc/massiveaction.class.php:1029                   MassiveAction->processForSeveralItemtypes()
  front/massiveaction.php:59                         MassiveAction->process()
  {"user":"2@3e251b604a39"} 
```